### PR TITLE
CDAP-3479- Integrate error dataset transformation - ETL Realtime

### PR DESCRIPTION
**Pull request by gsps1**
we want to user to be able to write to an error dataset from validator transform for invalid records. 
this PR contains implementation for ETL Realtime. 

JIRA: https://issues.cask.co/browse/CDAP-3479
Build: http://builds.cask.co/browse/CDAP-DUT2735-1

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-realtime-app/pom.xml [hunk 1]
```diff
@@ -52,6 +52,11 @@
       <version>${project.version}</version>
     </dependency>
     <dependency>
+      <groupId>co.cask.cdap</groupId>
+      <artifactId>cdap-etl-common</artifactId>
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-realtime-app/pom.xml]: we want to be able to use ``StructuredRecordStringConverter`` in ``ETLWorker``, so adding this dependency

File: cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactVersion.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ public String getSuffix() {
   }
 
   public boolean isSnapshot() {
-    return suffix != null && "snapshot".equals(suffix.toLowerCase());
+    return suffix != null && suffix.toLowerCase().contains("snapshot");
```
Comment (gsps1) [inline cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactVersion.java]: @albertshau let me know if we want to keep this.

Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactVersion.java]: i think we should, though it should probably be startsWith

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/DefaultEmitter.java [hunk 3]
```diff
@@ -56,6 +58,11 @@ public void emitError(InvalidEntry<T> value) {
     return entryList.iterator();
   }
 
+  public Iterator<InvalidEntry<T>> getErrorIterator() {
+    // we have to return iterator of a copy of list, as we need reset the errorList between stages.
+    return new ArrayList<>(errorList).iterator();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/DefaultEmitter.java]: does this mean we're not immediately consuming them?  the normal iterator() does not return a copy.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/ETLStage.java [hunk 4]
```diff
@@ -49,4 +55,8 @@ public String toString() {
       .add("properties", properties)
       .toString();
   }
+
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/ETLStage.java]: Nullable

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/ETLStage.java [hunk 5]
```diff
@@ -27,10 +27,16 @@
 public final class ETLStage {
   private final String name;
   private final Map<String, String> properties;
+  private final String datasetName;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/ETLStage.java]: this is a bit misleading. Perhaps name it errorDatasetName?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformDetails.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+package co.cask.cdap.template.etl.common;
+
+import javax.annotation.Nullable;
+
+/**
+ * Class to encapsulate id of the transform and the error dataset used in that transform stage.
+ */
+public class TransformDetails {
+  String transformId;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformDetails.java]: private final

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformDetails.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  public TransformDetails(String transformId, @Nullable String errorDatasetName) {
+    this.transformId = transformId;
+    this.errorDatasetName = errorDatasetName;
+  }
+
+  public String getTransformId() {
+    return transformId;
+  }
+
+  public String getErrorDatasetName() {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformDetails.java]: annotate as Nullable

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/PipelineRegisterer.java [hunk 8]
```diff
@@ -94,9 +106,18 @@ public Pipeline registerPlugins(ETLConfig config) {
         throw new IllegalArgumentException(String.format("No Plugin of type '%s' named '%s' was found",
                                                          Constants.Transform.PLUGINTYPE, transformConfig.getName()));
       }
+      // if the transformation is configured to write filtered records to error dataset, we create that dataset.
+      if (transformConfig.getDatasetName() != null) {
+        // TODO : can remove this after implementing CDAP-3480
+        if (errorDatasetType != null) {
+          configurer.createDataset(transformConfig.getDatasetName(), errorDatasetType, DatasetProperties.builder()
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/PipelineRegisterer.java]: perhaps we should also pass in the DatasetProperties?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/DefaultEmitter.java (hunk 3 above)
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/DefaultEmitter.java]: the normal iterator just updates a single list between each stage of the transform and after all the transform stages are complete, it just has to return the iterator of the list. whereas for our use case, we are interested in error records from each stage, so we have to do a copy. ``runOneIteration`` runs transform on all stages before returning.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformDetails.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package co.cask.cdap.template.etl.common;
+
+import javax.annotation.Nullable;
+
+/**
+ * Class to encapsulate id of the transform and the error dataset used in that transform stage.
+ */
+public class TransformDetails {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformDetails.java]: nit: lets rename to TransformDetail

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformDetails.java]: was confused for a little bit because this looks so similar to TransformationDetails.  Lets rename this to TransformInfo or something like that, and TransformationDetails to TransformDetail

File: cdap-app-templates/cdap-etl/cdap-etl-batch-app/src/main/java/co/cask/cdap/app/etl/batch/ETLMapReduce.java [hunk 10]
```diff
@@ -161,43 +162,40 @@ public void initialize(MapReduceContext context) throws Exception {
...
 
-      transformExecutor = new TransformExecutor<>(pipeline, stageMetrics);
+      transformExecutor = new TransformExecutor<>(pipeline);
     }
 
-    private void addTransforms(List<Transformation> pipeline,
-                               List<StageMetrics> stageMetrics,
-                               List<String> transformIds,
+    private void addTransforms(List<TransformationDetails> pipeline,
+                               List<TransformDetails> transformIds,
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch-app/src/main/java/co/cask/cdap/app/etl/batch/ETLMapReduce.java]: rename variable since its not longer ids

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformExecutor.java [hunk 11]
```diff
@@ -36,35 +36,45 @@
...
-    Transformation transform = transforms.get(0);
+    TransformationDetails transformationDetails = transforms.get(0);
     DefaultEmitter currentEmitter = emitters.get(0);
     currentEmitter.reset();
+    Transformation transform = transformationDetails.getTransformation();
     transform.transform(input, currentEmitter);
 
-    DefaultEmitter previousEmitter = currentEmitter;
+    // if there are errors in the iterator, we add it to the errorIterators list
+    if (currentEmitter.getErrorIterator().hasNext()) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformExecutor.java]: rather than tracking all these iterators, I think it would be cleaner to have the TransformExecutor take in an InvalidEntryWriter.  The InvalidEntryWriter writes any invalid entries it sees in between transform stages.  Then you don't need to introduce the TransformResponse object and the logic should be simpler.

File: cdap-app-templates/cdap-etl/cdap-etl-realtime-app/src/main/java/co/cask/cdap/app/etl/realtime/ETLWorker.java [hunk 12]
```diff
@@ -227,8 +241,31 @@ public void run(DatasetContext context) throws Exception {
...
+          }
+          final List<TransformResponse.TransformError> errorRecords = transformResponse.getErrorRecords();
+
+          for (final TransformResponse.TransformError response : errorRecords) {
+            final String datasetName = tranformIdToDatasetName.get(response.getTransformId());
+            getContext().execute(new TxRunnable() {
+              @Override
+              public void run(DatasetContext context) throws Exception {
+                Table errorTable = context.getDataset(datasetName);
+                String rowkeyPrefix = System.currentTimeMillis() + SEPARATOR + appName + SEPARATOR +
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-realtime-app/src/main/java/co/cask/cdap/app/etl/realtime/ETLWorker.java]: row key can be composed a little better to save bytes.  Timestamp is a long, followed by a string for app name, separator, transformid, then UUID is 128 bits.  Bytes has a method to interact with UUID.

File: cdap-app-templates/cdap-etl/cdap-etl-realtime-app/src/main/java/co/cask/cdap/app/etl/realtime/ETLWorker.java [hunk 13]
```diff
@@ -269,6 +306,16 @@ public void run(DatasetContext context) throws Exception {
     }
   }
 
+  private Put constructErrorPut(byte[] rowKey, InvalidEntry entry) throws IOException {
+    Put errorPut = new Put(rowKey);
+    errorPut.add(Constants.ErrorDataset.COLUMN_ERRCODE, entry.getErrorCode());
+    errorPut.add(Constants.ErrorDataset.COLUMN_ERRMSG, entry.getErrorMsg());
+    StructuredRecord record =  (StructuredRecord) entry.getInvalidRecord();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-realtime-app/src/main/java/co/cask/cdap/app/etl/realtime/ETLWorker.java]: nit: remove extract space before (StructuredRecord)

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-realtime-app/src/main/java/co/cask/cdap/app/etl/realtime/ETLWorker.java]: technically, there is no guarantee that the InvalidRecord is a StructuredRecord, even though this is true for all our transforms. Lets check that it really is a StructuredRecord, then just toString() it if it's not.

**Comment by albertshau** [comment]
took a pass through.  Mostly minor comments except for one about changing where we write to the error datasets

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformExecutor.java (hunk 11 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/TransformExecutor.java]: ok ignore that previous comment since we want to write everything at the same time in one transaction.  We can still make this better and avoid all the copying that is done by passing in a List<TransformError> to DefaultEmitter and just have the emitter add to the list.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/PipelineRegisterer.java (hunk 8 above)
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/PipelineRegisterer.java]: isnt the schema same between both of them? as we are writing InvalidEntry in both cases?

File: cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactVersion.java (hunk 2 above)
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactVersion.java]: actually lets remove this and do it in a separate PR.  It will exacerbate a bug we found today, such that master restarts will break plugins.

**Comment by gsps1** [comment]
@albertshau have addressed your comments, please take a look again. Thanks

**Comment by albertshau** [comment]
a few comments

**Comment by albertshau** [comment]
lgtm

**Comment by albertshau** [comment]
squashed commits and re-running build.
